### PR TITLE
[DT-405] Make deep links working when app is not started

### PR DESCRIPTION
**Pull request by G0retZ**
**What does this PR do?**

   The goal is to make deep links work if app is not started yet taking into account dependency on Home Screen data loading.

   Refactored the Home Screen logic in order to decouple it from the apps lists loading and use it for URLs cache initialisation.

**Database changed?**

   No

**Where should the reviewer start?**

- [x] I strongly recommend to review by commits to not get lost

**How should this be manually tested?**

  Flow on how to test this or QA Tickets related to this use-case: [APP-405](https://aptoide.atlassian.net/browse/APP-405)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APP-405](https://aptoide.atlassian.net/browse/APP-405)

**What are the relevant PRs?**

  PRs related to this one: [#135](https://github.com/Aptoide/aptoide-client-dt/pull/135)

**Code Review Checklist**

- [ ] Documentation on public interfaces
- [ ] Database changed? If yes - Migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] New Kotlin code has unit tests
- [ ] New flows in presenters unit tests
- [ ] Mappers/Validators with any kind of logic unit tests
- [ ] Functional tests pass

[APP-405]: https://aptoide.atlassian.net/browse/APP-405?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[APP-405]: https://aptoide.atlassian.net/browse/APP-405?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsListUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package cm.aptoide.pt.feature_apps.domain
+
+import cm.aptoide.pt.feature_apps.data.App
+
+interface AppsListUseCase {
+  suspend fun getAppsList(source: String): List<App>
+}
```
Comment (pedroagostinhoaptoide) [inline feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsListUseCase.kt:7]: Missing EOF

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/ESkillsAppsUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  private val urlsCache: UrlsCache
+) : AppsListUseCase {
+
+  override suspend fun getAppsList(source: String): List<App> =
+    appsRepository.getAppsList(
+      storeId = 15,
+      groupId = 14169744,
+      bypassCache = urlsCache.isInvalid(id = "eSkills/15/14169744")
+    )
+}
```
Comment (pedroagostinhoaptoide) [inline feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/ESkillsAppsUseCase.kt:21]: Missing EOF

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsByTagUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  override suspend fun getAppsList(source: String): List<App> = urlsCache.get(id = source)
+    ?.let {
+      appsRepository.getAppsList(
+        url = it,
+        bypassCache = urlsCache.isInvalid(id = source)
+      )
+    }
+    ?: throw IllegalStateException("No url cached")
+}
```
Comment (pedroagostinhoaptoide) [inline feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsByTagUseCase.kt:23]: Missing EOF

File: aptoide-network/src/main/java/cm/aptoide/pt/aptoide_network/domain/UrlsCache.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+interface UrlsCacheInitializer {
+  suspend fun initialise(): Map<String, String>
+}
+
+@JvmInline
+value class Blocker(private val channel: Channel<Unit> = Channel(0)) {
+  suspend fun await() = channel.receive()
+  fun yield() = channel.trySend(Unit)
+}
```
Comment (pedroagostinhoaptoide) [inline aptoide-network/src/main/java/cm/aptoide/pt/aptoide_network/domain/UrlsCache.kt:46]: Missing EOF

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/data/model/AppCategoryJSON.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package cm.aptoide.pt.feature_categories.data.model
+
+import androidx.annotation.Keep
+
+@Keep
+data class AppCategoryJSON(val name: String, val type: String)
```
Comment (pedroagostinhoaptoide) [inline feature_categories/src/main/java/cm/aptoide/pt/feature_categories/data/model/AppCategoryJSON.kt:6]: Missing EOF

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/AppCategory.kt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+package cm.aptoide.pt.feature_categories.domain
+
+data class AppCategory(val name: String, val type: String)
```
Comment (pedroagostinhoaptoide) [inline feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/AppCategory.kt:3]: Missing EOF

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/CategoriesUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+class CategoriesUseCase @Inject constructor(
+  private val urlsCache: UrlsCache,
+  private val categoriesRepository: CategoriesRepository,
+) {
+  suspend fun getCategories(tag: String): List<Category> =
+    urlsCache.get(id = tag)
+      ?.let { categoriesRepository.getCategoriesList(it) }
+      ?: throw IllegalStateException("No url found")
+}
```
Comment (pedroagostinhoaptoide) [inline feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/CategoriesUseCase.kt:15]: Missing EOF

File: feature_editorial/src/main/java/cm/aptoide/pt/feature_editorial/domain/usecase/ArticleUseCase.kt [hunk 8]
```diff
@@ -1,12 +1,19 @@
...
-class ArticleUseCase @Inject constructor(private val editorialRepository: EditorialRepository) {
+class ArticleUseCase @Inject constructor(
+  private val editorialRepository: EditorialRepository,
+  private val urlsCache: UrlsCache,
+) {
   suspend fun getDetails(articleId: String): Article =
-    editorialRepository.getArticle(articleId)
+    urlsCache.get(id = ARTICLE_CACHE_ID_PREFIX + articleId)
+      ?.let { editorialRepository.getArticle(editorialUrl = it) }
+      ?: throw IllegalStateException()
```
Comment (pedroagostinhoaptoide) [inline feature_editorial/src/main/java/cm/aptoide/pt/feature_editorial/domain/usecase/ArticleUseCase.kt:18]: Missing EOF

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/data/model/AppCategoryJSON.kt (hunk 5 above)
Comment (G0retZ) [reply feature_categories/src/main/java/cm/aptoide/pt/feature_categories/data/model/AppCategoryJSON.kt]: Fixed

File: feature_editorial/src/main/java/cm/aptoide/pt/feature_editorial/domain/usecase/ArticleUseCase.kt (hunk 8 above)
Comment (G0retZ) [reply feature_editorial/src/main/java/cm/aptoide/pt/feature_editorial/domain/usecase/ArticleUseCase.kt:18]: Fixed

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/CategoriesUseCase.kt (hunk 7 above)
Comment (G0retZ) [reply feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/CategoriesUseCase.kt]: Fixed

File: feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/AppCategory.kt (hunk 6 above)
Comment (G0retZ) [reply feature_categories/src/main/java/cm/aptoide/pt/feature_categories/domain/AppCategory.kt]: Fixed

File: aptoide-network/src/main/java/cm/aptoide/pt/aptoide_network/domain/UrlsCache.kt (hunk 4 above)
Comment (G0retZ) [reply aptoide-network/src/main/java/cm/aptoide/pt/aptoide_network/domain/UrlsCache.kt]: Fixed

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsByTagUseCase.kt (hunk 3 above)
Comment (G0retZ) [reply feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsByTagUseCase.kt]: Fixed

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/ESkillsAppsUseCase.kt (hunk 2 above)
Comment (G0retZ) [reply feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/ESkillsAppsUseCase.kt]: Fixed

File: feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsListUseCase.kt (hunk 1 above)
Comment (G0retZ) [reply feature_apps/src/main/java/cm/aptoide/pt/feature_apps/domain/AppsListUseCase.kt]: Fixed

**Pull request merged**